### PR TITLE
feat: spread group folder expiry background job

### PR DESCRIPTION
**Pull request by skjnldsv**

### CODE REVIEW

**Review by artonge — APPROVED** [review]

**Review by susnux — APPROVED** [review]

File: lib/BackgroundJob/ExpireGroupVersions.php [hunk 1]
```diff
@@ -42,7 +50,42 @@ public function __construct(GroupVersionsExpireManager $expireManager, ITimeFact
...
+		// If there are no folders, we don't need to do anything
+		if ($folderCount === 0) {
+			$this->logger->debug('No folders to expire', ['app' => 'cron']);
+			return;
+		}
+
+		// If we would go over the end of the list, wrap around
+		if ($lastFolder + $toDo >= $folderCount) {
+			$lastFolder = 0;
+		}
```
Comment (artonge) [inline lib/BackgroundJob/ExpireGroupVersions.php:76]: If `true`, then anything between `$lastFolder` and `$folderCount` will not be expired, no?

Comment (skjnldsv) [reply lib/BackgroundJob/ExpireGroupVersions.php:76]: Hmm, good point, let me adjust the logic

Comment (skjnldsv) [reply lib/BackgroundJob/ExpireGroupVersions.php]: Done

File: lib/BackgroundJob/ExpireGroupVersions.php [hunk 2]
```diff
@@ -24,14 +24,22 @@
...
 use OCP\BackgroundJob\TimedJob;
+use OCP\IAppConfig;
+use Psr\Log\LoggerInterface;
 
 class ExpireGroupVersions extends TimedJob {
-	private GroupVersionsExpireManager $expireManager;
-
-	public function __construct(GroupVersionsExpireManager $expireManager, ITimeFactory $timeFactory) {
+	public function __construct(
+		private ITimeFactory $timeFactory,
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply lib/BackgroundJob/ExpireGroupVersions.php]: Adjusted to use `protected` like the parent do

[the remark this answers is not in the document]
Comment (come-nc) [reply lib/BackgroundJob/ExpireGroupVersions.php]: As Ferdinand explained, you should not redeclare the property, it’s already defined in the parent class. The constructor property promotion feature adds a property to the class, which we do not want to do here.

File: lib/BackgroundJob/ExpireGroupVersions.php [hunk 3]
```diff
@@ -24,14 +24,22 @@
...
 use OCP\BackgroundJob\TimedJob;
+use OCP\IAppConfig;
+use Psr\Log\LoggerInterface;
 
 class ExpireGroupVersions extends TimedJob {
-	private GroupVersionsExpireManager $expireManager;
-
-	public function __construct(GroupVersionsExpireManager $expireManager, ITimeFactory $timeFactory) {
+	public function __construct(
+		protected ITimeFactory $timeFactory,
```
[the remark this answers is not in the document]
Comment (skjnldsv) [reply lib/BackgroundJob/ExpireGroupVersions.php]: Ah, sure!

[the remark this answers is not in the document]
Comment (skjnldsv) [reply lib/BackgroundJob/ExpireGroupVersions.php]: Well well well
```
Error: lib/BackgroundJob/ExpireGroupVersions.php:62:36: UndefinedThisPropertyFetch: Instance property OCA\GroupFolders\BackgroundJob\ExpireGroupVersions::$timeFactory is not defined (see https://psalm.dev/041)
```

**Review by artonge — APPROVED** [review]

File: lib/BackgroundJob/ExpireGroupVersions.php [hunk 4]
```diff
@@ -42,7 +50,42 @@ public function __construct(GroupVersionsExpireManager $expireManager, ITimeFact
...
+	 * Expiring groupfolder versions can be quite expensive.
+	 * We need to limit the amount of folders we expire per run.
+	 */
 	protected function run($argument) {
-		$this->expireManager->expireAll();
+		$lastFolder = $this->appConfig->getValueInt(Application::APP_ID, 'cron_last_folder_index', 0);
+		$folders = $this->folderManager->getAllFolders();
+
+		$folderCount = count($folders);
+		$currentRunHour = (int)date('G', $this->timeFactory->getTime());
```
Comment (come-nc) [inline lib/BackgroundJob/ExpireGroupVersions.php:62]:
```suggestion
		$currentRunHour = (int)date('G', $this->time->getTime());
```
It’s named `$time` in `Job`.

File: lib/BackgroundJob/ExpireGroupVersions.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (come-nc) [reply lib/BackgroundJob/ExpireGroupVersions.php]: See https://github.com/nextcloud/groupfolders/pull/3030#pullrequestreview-2163215822 , the property is named time, not timeFactory.

[the remark this answers is not in the document]
Comment (skjnldsv) [reply lib/BackgroundJob/ExpireGroupVersions.php]: Ah, thanks

**Review by come-nc — APPROVED** [review]

**Pull request merged**